ophyd/controls/device.py

```diff
@@ -0,0 +1,439 @@
+            raise ValueError('Must specify prefix if device signals are being '
+                             'used')
+
+        if name is None:
+            name = prefix
+
+        OphydObject.__init__(self, name=name, parent=parent)
+
+        if read_signals is None:
+            read_signals = self.signal_names
+
+        self.read_signals = list(read_signals)
+
+        # Instantiate non-lazy signals
+        [getattr(self, attr) for cpt, attr in self._sig_attrs.items()
+         if not cpt.lazy]
+
+    def wait_for_connection(self, all_signals=False, timeout=2.0):
+        '''Wait for signals to connect
+
+        Parameters
+        ----------
+        all_signals : bool, optional
+            Wait for all signals to connect (including lazy ones)
+        timeout : float or None
+            Overall timeout
+        '''
+        names = [attr for cpt, attr in self._sig_attrs.items()
+                 if not cpt.lazy or all_signals]
+
+        # Instantiate first to kickoff connection process
+        signals = [getattr(self, name) for name in names]
+
+        t0 = time.time()
+        while timeout is None or (time.time() - t0) < timeout:
+            connected = [sig.connected for sig in signals]
+            if all(connected):
+                return
+            time.sleep(min((0.05, timeout / 10.0)))
+
+        unconnected = [sig.name for sig in signals
+                       if not sig.connected]
+
+        raise TimeoutError('Failed to connect to all signals: {}'
+                           ''.format(', '.join(unconnected)))
+
+    @property
+    def connected(self):
+        return all(signal.connected for name, signal in self._signals.items())
+
+    def _get_devattr(self, name):
+        '''Gets a device attribute which may come from a sub-device'''
+        try:
+            owner_attr = self._sig_owner[name]
+        except KeyError:
+            raise ValueError('Unknown read signal: {}'.format(name))
+
+        if owner_attr is None:
+            # None means this instance owns it
+            return getattr(self, name)
+        else:
+            # Otherwise, get the owner first
+            owner = getattr(self, owner_attr)
+            # Then the attribute
+            return getattr(owner, name)
+
+    def read(self):
+        # map names ("data keys") to actual values
+        values = OrderedDict()
+        for name in self.read_signals:
+            signal = self._get_devattr(name)
+            values.update(signal.read())
+
+        return values
+
+    def describe(self):
+        desc = OrderedDict()
+        for name in self.read_signals:
+            signal = self._get_devattr(name)
+            desc.update(signal.describe())
+
+        return desc
+
+    @property
+    def trigger_signals(self):
+        names = [attr for cpt, attr in self._sig_attrs.items()
+                 if cpt.trigger_value is not None]
+
+        return [getattr(self, name) for name in names]
+
+    def trigger(self, **kwargs):
```

Comment:
Right. Perhaps they should be passed along to `put`, but as in #171 we should step back from looking at this particular implementation to seeing exactly what we want from it to start with...